drivers/adc/ad7091r8/ad7091r8.c

```diff
@@ -0,0 +1,609 @@
+}
+
+/**
+ * Write to device.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_data - The register data.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad7091r8_spi_reg_write(struct ad7091r8_dev *dev,
+			   uint8_t reg_addr,
+			   uint16_t reg_data)
+{
+	uint8_t buf[2];
+
+	if (!dev || !reg_data)
+		return -EINVAL;
+
+	/*
+	 * AD7091R-2/-4/-8 protocol (datasheet page 31) is to do a single SPI
+	 * transfer with reg address set in bits B15:B11 and value set in B9:B0.
+	 */
+	no_os_put_unaligned_be16(
+		no_os_field_prep(AD7091R8_REG_DATA_MSK, reg_data) |
+		no_os_field_prep(AD7091R8_RD_WR_FLAG_MSK, 1) |
+		no_os_field_prep(AD7091R8_REG_ADDR_MSK, reg_addr), buf);
+
+	return no_os_spi_write_and_read(dev->spi_desc, buf, 2);
+}
+
+/**
+ * Read from device.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_data - The register data.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad7091r8_spi_reg_read(struct ad7091r8_dev *dev,
+			  uint8_t reg_addr,
+			  uint16_t *reg_data)
+{
+	uint8_t buf[2];
+	int ret;
+
+	if (!dev || !reg_data)
+		return -EINVAL;
+
+	if (reg_addr == AD7091R8_REG_RESULT) {
+		ret = ad7091r8_pulse_convst(dev);
+		if (ret)
+			return ret;
+	}
+
+	no_os_put_unaligned_be16(
+		no_os_field_prep(AD7091R8_RD_WR_FLAG_MSK, 0) |
+		no_os_field_prep(AD7091R8_REG_ADDR_MSK, reg_addr), buf);
+
+	/* Reg data only comes out on the next transfer (datasheet figure 52) */
+	ret = no_os_spi_write_and_read(dev->spi_desc, buf, 2);
+	if (ret)
+		return ret;
+
+	if (reg_addr == AD7091R8_REG_RESULT) {
+		ret = ad7091r8_pulse_convst(dev);
+		if (ret)
+			return ret;
+	}
+
+	no_os_put_unaligned_be16(
+		no_os_field_prep(AD7091R8_RD_WR_FLAG_MSK, 0) |
+		no_os_field_prep(AD7091R8_REG_ADDR_MSK, reg_addr), buf);
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buf, 2);
+	if (ret)
+		return ret;
+
+	*reg_data = no_os_get_unaligned_be16(buf);
+	return 0;
+}
+
+/**
+ * SPI write to device using a mask.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param mask - The mask.
+ * @param data - The register data.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad7091r8_spi_write_mask(struct ad7091r8_dev *dev,
+			    uint8_t reg_addr,
+			    uint16_t mask,
+			    uint16_t data)
+{
+	uint16_t reg_data;
+	int ret;
+
+	if (!dev)
+		return -EINVAL;
+
+	ret = ad7091r8_spi_reg_read(dev, reg_addr, &reg_data);
+	if (ret)
+		return ret;
+
+	reg_data &= ~mask;
+	reg_data |= data;
+
+	return ad7091r8_spi_reg_write(dev, reg_addr, reg_data);
+}
+
+/**
+ * @brief Set device sleep mode.
+ *
+ * @param dev - The device structure.
+ * @param mode - The device sleep mode to set.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad7091r8_set_sleep_mode(struct ad7091r8_dev *dev,
+			    enum ad7091r8_sleep_mode mode)
+{
+	if (!dev)
```

Comment:
this check is useless here, dev is not used in this function, it's only passed to ad7091r8_spi_write_mask which will anyway return error if dev is NULL